aarch64: Math functions Implementations in C

thiru-mcw opened the pull request:
This PR introduces Math-Functions implementations for the aarch64 target, addressing [Cygwin: Port of mathematical functions](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/issues/251)

- Most functions are implemented in C to enable intial functionality, with a plan to transition them into arm64 assembly in a future update.
- For assembly files they are branched and linked to the fallback C implementations in the Newlib repository.
- Functions like square root utilize the existing opcode `sqrt` as they are already supported in aarch64 for efficient implementation.

Doc: [Math_funcs_test_programs.xlsx](https://github.com/user-attachments/files/19627526/Math_funcs_test_programs.xlsx)

thiru-mcw commented:
This PR addresses the issue mentioned in [Cygwin: Port mathematical functions to AArch64](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/issues/251) .

Blackhex commented on winsup/cygwin/math/cosl_internal.S:
```diff
@@ -51,7 +51,12 @@ __MINGW_USYMBOL(__cosl_internal):
+  ret
 #else
-  // TODO
+  .error "Not supported on your platform yet"
```
Do I assume correctly that this implementation is not used anywhere we know so far? The same question applies to `exp2l.S`.

thiru-mcw replied:
I added this `#else` block to prevent unsupported platforms from compiling, making the lack of implementation clear to users. 
It can be reverted if not required, same as in the case of `exp2l.S`.

Blackhex commented on winsup/cygwin/math/exp.def.h:
```diff
@@ -103,8 +103,12 @@ __expl_internal (long double x)
+  long double i;
+  long double f = modfl(y, &i);
+  long double z = 1.0L + f + (c1 * f);
+  res=ldexpl(z, (int)i);
```
Please keep the code style consistent with GNU standards, I believe it should be:
```
res = ldexpl(z, (int) i);
```

Blackhex commented on winsup/cygwin/math/frexpl.S:
```diff
@@ -129,6 +129,12 @@ L24:
 	ret
 #elif defined(__aarch64__)
 	// TODO
```
Please update the `TODO` with message explaining that full asembly implementation is missing.

thiru-mcw replied:
updating with // TODO: Complete AArch64 assembly implementation

Blackhex reviewed (approved):
I can approve as an icremental improvement of the current state. Left some minor comments. Proper assembly implementation would be needed in the future.

Blackhex commented on winsup/cygwin/math/fastmath.h:
```diff
@@ -116,8 +115,8 @@ static __inline__ long double __fast_log1pl (long double x)
 #elif defined(__aarch64__)
-    // TODO
-    res = 0.0;
+      // TODO: Complete AArch64 assembly implementation
```
Here should be only 4 spaces indentation.

thiru-mcw replied:
Thanks for pointing out that, will fix in next push.

thiru-mcw replied:
Done

Blackhex commented on winsup/cygwin/math/log10l.S:
```diff
@@ -34,7 +34,7 @@ __MINGW_USYMBOL(log10l):
 	fld	%st		// x : x : log10(2)
 	sahf
-	jc	3f		// in case x is NaN or �Inf
+	jc	3f		// in case x is NaN or �Inf
```
Please revert. The file is not utf-8 encoded and should be edited in its original encoding. The same applies to other files.

thiru-mcw replied:
Done, Thanks for pointing out that, these files get affected automatically when touched.

Blackhex reviewed (changes requested):
There are some files encoding issues. Otherwise look good as an iterative improvement.

thiru-mcw commented:
All the 31 Functions are being ported, and functionally tested with sample programs.
Separate commits were squashed into a single commit: **[Adding math-function to winsup](https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/11/commits/ba53478577a225509514407f84200a05f2431747)**

eukarpov commented on winsup/cygwin/math/cosl_internal.S:
```diff
@@ -51,7 +51,13 @@ __MINGW_USYMBOL(__cosl_internal):
+	ret
 #else
-  // TODO
+  .error "Not supported on your platform yet"
```
I would suggest to replace with something like
"Not supported architecture yet"

thiru-mcw replied:
Done, applied the suggested change.

eukarpov reviewed (approved):
LGTM as an intermediate step.
How this implementation has been tested?

thiru-mcw committed:
Adding math-function to winsup

atan2l: Add C implementation

atanl: Add C implementation

asinl: Add C implementation

cossin: Add C Implementations

Adding C implementation for sincos, sincosf, sincosl.

sqrt.def.h: Adding asm Implementation

Same opcode `sqrt` is available in aarch64, so directly implemented it alike to x64.

Remainder: Adding asm Implementation

fmodl: Adding C implementation

ldexpl: Add C Implementation

pow.def.h: Adding C implementation

Add c implementation for internal_modf & asm implementation for fsqrt

logbl: Adding C implementation

exp.def.h: Add C Implementation

tanl: Adding asm implementation

scalbln: Add asm implementation

ilogbl: Add asm implementaion

exp2l: Add asm implementations

frexpl: Adding asm implementation

remquol: Add asm Implementation

expm1.def.h: Add C implementation

cosl_internal: Add asm implementation

internal_logl: Add asm Implementation

sinl_internal: Add asm Implementation

fastmath.h: Add C implementation

log1pl: Add asm Implementation

log2l: Add asm Implementation

log10l: Add asm implementation

scalbl: Add asm Implementation

thiru-mcw commented:
> How this implementation has been tested?

@eukarpov After each implementation, a sample test program (as mentioned in this document: [Math_funcs_test_programs.xlsx](https://github.com/user-attachments/files/19627710/Math_funcs_test_programs.xlsx) ) was used for verification. The results were compared against outputs from the x86 toolchain and online compilers to ensure correctness.

Note: For the assembly-based implementation, correct behavior is observed only when compiled with the -O2 optimization flag.
